header-rewriter/GenCallAsm.cpp

```diff
@@ -216,6 +216,72 @@ static void emit_mixed_wrpkru(AsmWriter &aw, const std::string &pkey0,
   add_raw_line(aw, "IA2_WRPKRU \"\\n\"");
 }
 
+static void emit_load_pkey(AsmWriter &aw, const std::string &pkey,
+                           const std::string &reg) {
+  add_asm_line(aw, llvm::formatv("mov $\" XSTR({0}) \", %{1}", pkey, reg));
+}
+
+// Emit code to load the address of a compartment's stack from ia2_stackptrs.
+// \p sp_addr_reg and \p compartment_offset_reg are register names (sans %
+// prefix); after this asm runs, sp_addr_reg will hold the address where the
+// compartment's stack pointer is stored, and compartment_offset_reg will hold
+// a compartment-specific offset.
+static void emit_load_sp_addr(AsmWriter &aw, const std::string &pkey,
+                              const std::string &sp_addr_reg,
+                              const std::string &compartment_offset_reg) {
+  // Compute compartment offset in compartment_offset_reg
+  emit_load_pkey(aw, pkey, compartment_offset_reg);
+  add_asm_line(aw, llvm::formatv("shl $12, %{0}", compartment_offset_reg));
```

Comment:
Guessing `$12` here is the page size? I wonder if we want to move the constant from the output assembly text into the C++ code, and maybe even compute it at run time (or maybe as an IA2 build constant).

Reply:
This runtime computation of the page offset will be switched to a codegen-time constant when #182 gets merged since that PR changes the caller/target pkeys in emit_asm_wrapper to known integers instead of macros. So I think it's probably fine as-is for now.